File: java/src/main/java/nl/vermeir/java/service/CustomerService.java
package nl.vermeir.java.service;

import nl.vermeir.java.model.Customer;
import nl.vermeir.java.repository.CustomerRepository;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@Service
public class CustomerService {

    private final CustomerRepository customerRepository;
    CustomerService(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    public Customer addCustomer(Customer customer) {
        return customerRepository.save(customer);
    }

    public Iterable<Customer> getAllCustomers() {
        return customerRepository.findAll();
    }

    public Optional<Customer> findCustomerById(String id) {
        return customerRepository.findById(id);
    }

    public void deleteCustomer(String id) {
        customerRepository.deleteById(id);
    }

    public Customer updateCustomer(Customer customer) {
        return customerRepository.save(customer);
    }
}

File: kotlin/src/main/kotlin/nl/vermeir/kotlin/service/CustomerService.kt
package nl.vermeir.kotlin.service

import nl.vermeir.kotlin.controller.Customer
import nl.vermeir.kotlin.repository.CustomerAlreadyExistsException
import nl.vermeir.kotlin.repository.CustomerRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import java.util.*

@Service
class CustomerService(val db: CustomerRepository) {

    fun findCustomers(): List<Customer> = db.findAll().toList()

    fun findCustomerById(id: String): Customer = db.findById(id).orElseThrow()

    fun saveCustomer(customer: Customer): Customer {
        try {
            return db.save(customer)
        } catch (e: DataIntegrityViolationException) {
            throw CustomerAlreadyExistsException("Customer with the same name or email already exists.")
        } catch (e: Exception) {
            throw e
        }
    }

    fun deleteCustomer(id: String) {
        db.deleteById(id)
    }

    fun <T : Any> Optional<out T>.toList(): List<T> = if (isPresent) listOf(get()) else emptyList()

}